bpo-40503: Add compile-time configuration to PEP 615

pganssle opened the pull request:
This is the implementation for PEP 615's support for compile-time configuration of the time zone search path.

The main implementation is in #19909 — at the moment it's not merged so unfortunately the commits are also present in this PR until that one gets merged. I couldn't come up with a good way to both have a neat diff and run the CI, so I went with running the CI.

The relevant changes are the ones starting with the commit "`Add --with-tzpath to autoconf`"

<!-- issue-number: [bpo-40503](https://bugs.python.org/issue40503) -->
https://bugs.python.org/issue40503
<!-- /issue-number -->

pganssle committed:
Add tests and implementation for ZoneInfo

pganssle committed:
Add tzdata to Travis requirements

pganssle committed:
Deliberately break TZData tests to see who is skipping them

pganssle committed:
Add test requirements to Github Actions builds

pganssle committed:
Add test dependencies to windows AP builds

pganssle committed:
Revert "Add test dependencies to windows AP builds"

This reverts commit b5de7784fa7e3e1d1fc2de97932f52bafbacf8d2.

pganssle committed:
Revert "Deliberately break TZData tests to see who is skipping them"

This reverts commit 45146f1a12a4fb393fa5ea100be01e1cb6d45d30.

pganssle committed:
fixup! Add tests and implementation for ZoneInfo

pganssle committed:
fixup! Add tests and implementation for ZoneInfo

pganssle committed:
fixup! Add tests and implementation for ZoneInfo

pganssle committed:
fixup! Add tests and implementation for ZoneInfo

pganssle committed:
fixup! fixup! Add tests and implementation for ZoneInfo

pganssle committed:
Add --with-tzpath to autoconf

This is configurable only on POSIX systems at the moment and TZPATH is
initialized to an empty string on Windows.

Yhg1s commented on Makefile.pre.in:
```diff
@@ -197,6 +197,9 @@ OPENSSL_INCLUDES=@OPENSSL_INCLUDES@
 OPENSSL_LIBS=@OPENSSL_LIBS@
 OPENSSL_LDFLAGS=@OPENSSL_LDFLAGS@
 
+# zoneinfo.TZPATH default
```
Add a note that this isn't used anywhere in the Makefile, and that it's just there for sysconfig.get_config_var().

Yhg1s commented on configure.ac:
```diff
@@ -2946,6 +2946,44 @@ esac
+validate_tzpath() {
+    # Check that each element of the tzpath is an absolute path
+    all_paths=$1
+    while test "${all_paths#*:}" != "$1"; do
```
The comparison to "$1" confuses me. $all_paths starts out as $1, and you're only removing things, so surely this loop would never end? It also feels like you could do this with grep instead:

  if ( echo $1 | grep -qE '(^|:)([^/]|$)' ); then
      AC_MSG_ERROR(...);
      return 1;
  fi

(There are already other uses of grep -E in configure.ac, so I think it's safe to rely on.)

pganssle replied:
I like it, though it was annoyingly tricky to pull off. It took me *forever* to realize that `autoconf` was stripping out the `[]` characters from the grep string.

Yhg1s commented on configure.ac:
```diff
@@ -2946,6 +2946,44 @@ esac
+    done
+}
+
+TZPATH="/usr/share/zoneinfo:/usr/lib/zoneinfo:/usr/share/lib/zoneinfo:/etc/zoneinfo"
```
This should be passed to validate_tzpath as well.

Yhg1s commented on configure.ac:
```diff
@@ -2946,6 +2946,44 @@ esac
+        ;;
+esac
+],
+[AC_MSG_RESULT(default)])
```
Add $TZPATH in the message here, so you can see what that default is?

pganssle replied:
Good catch. I originally had it this way, but then I played around with the possibility of using something closer to the previous logic, where if `TZPATH` is not explicitly set, at runtime the module will order the path based on a hard-coded string where the order is determined by which directories exist at import time. In that case, "default" made more sense than any hard-coded string.

I abandoned that approach as YAGNI — if we want to do something like that, we should probably just do it at build time anyway.

Yhg1s commented on setup.py:
```diff
@@ -304,6 +304,19 @@ def find_library_file(compiler, libname, std_dirs, paths):
+    tzpaths = base_tzpath.split(os.pathsep)
+    bad_paths = [tzpath for tzpath in tzpaths if not os.path.isabs(tzpath)]
+    if bad_paths:
+        raise ValueError('TZPATH must contain only absolute paths, '
```
Use a format string with {!r} instead?

pganssle committed:
Fix member def for key

pganssle committed:
Fix refleak in error code

pganssle committed:
Fix refleak on cache miss

pganssle committed:
fixup! Add tests and implementation for ZoneInfo

pganssle committed:
Fix memory leak in zoneinfo_dealloc

pganssle committed:
Fix refleak in fromutc

pganssle committed:
Add smoke test that C and Python modules built

The import machinery can be somewhat fragile, and the "seamlessly falls
back to pure Python" nature of this module makes it so that a problem
building the C extension or a failure to import the pure Python version
might easily go unnoticed.

This adds some "smoke tests" that rely on implementation details of each
module to ensure that we're building the ones we think we are.

pganssle committed:
Add Windows build support

This should improve support on Windows, including actually building the C
extension.

pganssle committed:
fixup! Add Windows build support

pganssle committed:
Revert "fixup! Add Windows build support"

This reverts commit 874dca0a79c76b668cf68f665c8490af53306c85.

pganssle committed:
fixup! Add Windows build support

Yhg1s commented on configure.ac:
```diff
@@ -2946,6 +2946,42 @@ esac
 
+validate_tzpath() {
+    # Checks that each element of hte path is an absolute path
+    if test "$1" == ""; then
```
if test -z "$1"; then

Yhg1s commented on configure.ac:
```diff
@@ -2946,6 +2946,42 @@ esac
+TZPATH="/usr/share/zoneinfo:/usr/lib/zoneinfo:/usr/share/lib/zoneinfo:/etc/zoneinfo"
+AC_MSG_CHECKING(for --with-tzpath)
+AC_ARG_WITH(tzpath,
+            AS_HELP_STRING([--with-tzpath=<list of relative paths separated by pathsep>]
```
Absolute paths, yes?

pganssle replied:
Hah, oops! Good catch.

Yhg1s reviewed (approved):
LG for the configure changes (for UNIX; not sure if more should be done for Windows).

pganssle committed:
Remove __version__ from __dir__

pganssle committed:
Include all globals in __dir__

Suggestion by Petr Viktorin.

zooba commented on .github/workflows/build.yml:
```diff
@@ -25,6 +25,10 @@ jobs:
+    - name: Install test dependencies
+      run: |
+        .\python.bat -m ensurepip --user
+        .\python.bat -m pip install --user -r Misc/requirements-test.txt
```
As we discussed, I don't like relying on pip to install into the source tree before we've run the test suite.

In general, system-wide dependencies are installed using the system package manager, or downloaded using specialised scripts and an existing Python installation (see `PCbuild/get-externals.py`). This would be a new network dependency, and I'd rather keep us to things that are checked in or directly grabbed from GitHub.

zooba commented on .github/workflows/coverage.yml:
```diff
@@ -48,6 +48,7 @@ jobs:
         ./python -m venv .venv
         source ./.venv/bin/activate
         python -m pip install -U coverage
+        python -m pip install -r Misc/requirements-test.txt
```
Also as discussed, coverage is the one exception, but it's also not a core part of the PR workflow. We already switch CI providers for PRs every time Victor gets annoyed at network instability, so let's try and keep PR clean. We can leave it in the post-merge workflow if you really want.

zooba commented on Lib/sysconfig.py:
```diff
@@ -546,6 +546,7 @@ def get_config_vars(*args):
 
         if os.name == 'nt':
             _init_non_posix(_CONFIG_VARS)
+            _CONFIG_VARS['TZPATH'] = ''
```
Shouldn't we just `.get()` it rather than assuming it'll be set?

pganssle replied:
I think we should be guaranteeing that it's set, and it should be an error condition otherwise. The "reasonable default" varies by platform, and can be configured at build time.

On Windows, the reasonable default is actually an empty string (no tzdata), but also I'd like to add it as a flag at build time.

zooba commented on Lib/test/test_zoneinfo/_support.py:
```diff
@@ -0,0 +1,76 @@
+TZPATH_TEST_LOCK = threading.Lock()
+
+
+def call_once(f):
```
:trollface:

zooba commented on Lib/test/test_zoneinfo/test_zoneinfo.py:
```diff
@@ -0,0 +1,1994 @@
+        subclass = [False, True]
+
+        key = "Europe/London"
+        zi = self.zone_from_key("Europe/London")
```
```suggestion
        zi = self.zone_from_key(key)
```

zooba commented on Lib/zoneinfo/__init__.py:
```diff
@@ -0,0 +1,29 @@
+from ._common import ZoneInfoNotFoundError
+
+try:
+    from _czoneinfo import ZoneInfo
```
We normally just use the underscore prefix to indicate an accelerator module. The `c` prefix hasn't really been used since 3.0

pganssle replied:
This is a slightly more complicated situation, because `_czoneinfo` is actually parallel to `zoneinfo._zoneinfo`. In the reference implementation, the two are `zoneinfo._zoneinfo` and `zoneinfo._czoneinfo`, but the limitations of our current build system prevent me from doing that here.

zooba commented on Lib/zoneinfo/_tzpath.py:
```diff
@@ -0,0 +1,110 @@
+def _parse_python_tzpath(env_var):
+    if not env_var:
+        return ()
+    else:
```
Unnecessary `else`, and then you can dedent the entire the function from here on.

pganssle replied:
Good point, thanks. Resolving in the primary PR for this #19909.

zooba commented on Lib/zoneinfo/_zoneinfo.py:
```diff
@@ -0,0 +1,756 @@
+            + day
+        )
+
+    # TODO: These are not actually epoch dates as they are expressed in local time
```
Does this TODO need to be fixed?

pganssle replied:
At some point, but this is a method on a private class and the only issue is that the name is slightly misleading, so it's not in the critical path to an MVP.

zooba commented on PCbuild/_czoneinfo.vcxproj:
```diff
@@ -0,0 +1,109 @@
+    </ProjectConfiguration>
+  </ItemGroup>
+  <PropertyGroup Label="Globals">
+    <ProjectGuid>{384C224A-7474-476E-A01B-750EA7DE918C}</ProjectGuid>
```
This GUID should be replaced with a new one (similarly in `PCbuild.sln`)

Here you go:
```suggestion
    <ProjectGuid>{FCBE1EF2-E0F0-40B1-88B5-00A35D378742}</ProjectGuid>
```

zooba commented on PCbuild/pcbuild.sln:
```diff
@@ -91,6 +91,8 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "_testconsole", "_testconsol
 EndProject
 Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "_asyncio", "_asyncio.vcxproj", "{384C224A-7474-476E-A01B-750EA7DE918C}"
 EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "_czoneinfo", "_czoneinfo.vcxproj", "{384C224A-7474-476E-A01B-750EA7DE918C}"
```
```suggestion
Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "_czoneinfo", "_czoneinfo.vcxproj", "{FCBE1EF2-E0F0-40B1-88B5-00A35D378742}"
```

zooba reviewed (commented):
Windows build changes look fine, thanks for getting the installer!

pganssle committed:
fixup! Add Windows build support

pganssle committed:
Reduce indentation in _parse_python_tzpath
